### PR TITLE
ddl(ticdc): allow to rename tables

### LINKED ISSUE #3822: rename multiple tables in one statement not sync to downstream
### What did you do?

I need to backup data to new table, so I create a new table, and exchange the names of both. but this action not sync to downstream TiDB.
```
rename table pre_asset_ca_gbc_fof to bak_pre_asset_ca_gbc_fof_20211209,new_pre_asset_ca_gbc_fof to pre_asset_ca_gbc_fof
```
and some information in logs
```
[2021/12/09 14:13:45.147 +08:00] [INFO] [schema_storage.go:832] ["discard the ddl job"] [jobID=3302] [query="rename table pre_asset_ca_gbc_fof to bak_pre_asset_ca_gbc_fof_20211209,new_pre_asset_ca_gbc_fof to pre_asset_ca_gbc_fof"]
```
So, I search in the source code, and find this in the function `pkg/filter/filter.go:127: Filter.shouldDiscardByBuiltInDDLAllowlist`:
```golang
func (f *Filter) shouldDiscardByBuiltInDDLAllowlist(ddlType model.ActionType) bool {
	/* The following DDL will be filter:
	ActionAddForeignKey                 ActionType = 9
	ActionDropForeignKey                ActionType = 10
	ActionRebaseAutoID                  ActionType = 13
	ActionShardRowID                    ActionType = 16
	ActionLockTable                     ActionType = 27
	ActionUnlockTable                   ActionType = 28
	ActionRepairTable                   ActionType = 29
	ActionSetTiFlashReplica             ActionType = 30
	ActionUpdateTiFlashReplicaStatus    ActionType = 31
	ActionCreateSequence                ActionType = 34
	ActionAlterSequence                 ActionType = 35
	ActionDropSequence                  ActionType = 36
	ActionModifyTableAutoIdCache        ActionType = 39
	ActionRebaseAutoRandomBase          ActionType = 40
	ActionAlterIndexVisibility          ActionType = 41
	ActionExchangeTablePartition        ActionType = 42
	ActionAddCheckConstraint            ActionType = 43
	ActionDropCheckConstraint           ActionType = 44
	ActionAlterCheckConstraint          ActionType = 45
	ActionAlterTableAlterPartition      ActionType = 46

	... Any Action which of value is greater than 46 ...
	*/
	switch ddlType {
	case model.ActionCreateSchema,
		model.ActionDropSchema,
		model.ActionCreateTable,
		model.ActionDropTable,
		model.ActionAddColumn,
		model.ActionDropColumn,
		model.ActionAddIndex,
		model.ActionDropIndex,
		model.ActionTruncateTable,
		model.ActionModifyColumn,
		model.ActionRenameTable,
		model.ActionSetDefaultValue,
		model.ActionModifyTableComment,
		model.ActionRenameIndex,
		model.ActionAddTablePartition,
		model.ActionDropTablePartition,
		model.ActionCreateView,
		model.ActionModifyTableCharsetAndCollate,
		model.ActionTruncateTablePartition,
		model.ActionDropView,
		model.ActionRecoverTable,
		model.ActionModifySchemaCharsetAndCollate,
		model.ActionAddPrimaryKey,
		model.ActionDropPrimaryKey,
		model.ActionAddColumns,
		model.ActionDropColumns:
		return false
	}
	return true
}
```
What's more, I find ActionType defined in github.com/pingcap/parser/model/ddl.go, and there are two type of `rename  table`
```
	ActionRenameTable                   ActionType = 14
        ...
	ActionRenameTables                  ActionType = 47
```
So, just add `model.ActionRenameTables` to the function `Filter.shouldDiscardByBuiltInDDLAllowlist`, and the bug will be fixed

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
Release Version: v5.1.0
Edition: Community
Git Commit Hash: 8acd5c88471cb7b4d4c4a8ed73b4d53d6833f13e
Git Branch: heads/refs/tags/v5.1.0
UTC Build Time: 2021-06-24 07:10:32
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

TiCDC version (execute `cdc version`):

```console
Release Version: v5.1.0
Git Commit Hash: b5c7b35da4fe4fa934b1fdd51b51380f588f9ee5
Git Branch: heads/refs/tags/v5.1.0
UTC Build Time: 2021-06-24 07:09:35
Go Version: go version go1.16.4 linux/amd64
```

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve? 

Close #3822 .

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by overvenus — COMMENTED** [review]
Please also add an integration test, we can add it to `ddl_reentrant`

File: cdc/model/sink.go [hunk 1]
```diff
@@ -455,6 +455,13 @@ func (d *DDLEvent) FromJob(job *model.Job, preTableInfo *TableInfo) {
 	d.CommitTs = job.BinlogInfo.FinishedTS
 	d.Query = job.Query
 	d.Type = job.Type
+	d.fillPreTableInfo(preTableInfo)
+
+	switch d.Type {
+	case model.ActionRenameTables:
+		return
```
[the remark this answers is not in the document]
Comment (hicqu) [reply cdc/model/sink.go:464]: Updated.

File: cdc/entry/schema_storage.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ func (s *SingleSchemaSnapshot) PreTableInfo(job *timodel.Job) (*model.TableInfo,
 			return nil, cerror.ErrSchemaStorageTableMiss.GenWithStackByArgs(job.TableID)
 		}
 		return table, nil
+	case timodel.ActionRenameTables:
+		// DDL on multiple tables, ignore pre table info
```
Comment (leoppro) [inline cdc/entry/schema_storage.go:81]: Why the pre-table-info can be ignored?

File: cdc/redo/common/util.go [hunk 3]
```diff
@@ -49,7 +49,6 @@ var InitS3storage = func(ctx context.Context, uri url.URL) (storage.ExternalStor
 	s3storage, err := storage.New(ctx, backend, &storage.ExternalStorageOptions{
 		SendCredentials: false,
 		HTTPClient:      nil,
-		SkipCheckPath:   true,
```
Comment (hicqu) [inline cdc/redo/common/util.go:52]: @ben1009 please take a look here.

File: cdc/entry/schema_storage.go [hunk 4]
```diff
@@ -584,6 +591,36 @@ func (s *schemaSnapshot) handleDDL(job *timodel.Job) error {
...
+		for _, tableID := range oldTableIDs {
+			if err := s.dropTable(tableID); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for i, tableInfo := range job.BinlogInfo.MultipleTableInfos {
+			newSchema, ok := s.SchemaByID(newSchemaIDs[i])
+			if !ok {
+				err := errors.New("Can't find schema for rename tables")
+				return errors.Trace(err)
```
Comment (overvenus) [inline cdc/entry/schema_storage.go:615]:
```suggestion
				return cerror.ErrSnapshotSchemaNotFound.GenWithStackByArgs(newSchemaIDs[i])
```

File: cdc/entry/schema_storage.go [hunk 5]
```diff
@@ -584,6 +591,36 @@ func (s *schemaSnapshot) handleDDL(job *timodel.Job) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+	case timodel.ActionRenameTables:
+		var oldSchemaIDs, newSchemaIDs, oldTableIDs []int64
```
Comment (overvenus) [inline cdc/entry/schema_storage.go:595]: Could you add a new function `renameTables()`?

File: cdc/entry/schema_storage.go [hunk 6]
```diff
@@ -584,6 +591,36 @@ func (s *schemaSnapshot) handleDDL(job *timodel.Job) error {
...
+		var newTableNames, oldSchemaNames []*timodel.CIStr
+		err := job.DecodeArgs(&oldSchemaIDs, &newSchemaIDs, &newTableNames, &oldTableIDs, &oldSchemaNames)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(job.BinlogInfo.MultipleTableInfos) < len(newTableNames) {
+			err := errors.New("Invalid binlog info for rename tables")
+			return errors.Trace(err)
```
Comment (overvenus) [inline cdc/entry/schema_storage.go:603]: Could you add a new error in `pkg/errors`, e.g., `ErrInvalidDDLJob`?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Review by overvenus — APPROVED** [review]
LGTM

**Review by ben1009 — APPROVED** [review]

**Pull request merged**